### DISCUSSION TITLE
Omit optional fields in `validate_plugin_config_schema` result?

### DISCUSSION THREAD

**Discussion by jsumners** [Kong/kong#9097] [Help]
I have a schema that has several fields with default values and a few fields that are optional (no default values and don't have to be specified). When I use `require('spec.helpers').validate_plugin_config_schema` to validate the schema in a unit test, I want the optional fields to not be present in the validated schema. For example:

```lua
-- schema.lua
local schema = {
  {
    config = {
      type = 'record',
      fields = {
        { foo = { type = 'string', default = 'foo' } },
        { bar = { type = 'string' }
      }
    }
  }
}
```

```lua
local validated, err = validate_plugin_config_schema({}, schema)

-- what I want
assert.same({ foo = 'foo' }, validated)

-- what I get
assert.same({ foo = 'foo', bar = 'userdata: NULL' }, validated)
```

If I attach `abstract = true` to the `bar` field definition then it won't show up in the `validated` result. But, if I do that, then when I try to provide a value for the `bar` field I get some "abstract field was not specialized" error.

How can I accomplish my goal? What is this `abstract` property?

**Comment by hanshuebner** [comment]
Hello @jsumners,

thank you for your question.  Can you actually change your test to compare the validation result field by field, rather than comparing the whole table?  That will also be more robust should additional fields be added to the validation results.

The string `userdata: NULL` is actually a `nil` and it'd be nicer if it showed up in the validation result as `nil` - That would be somewhat of a bug that might warrant a fix.

The `abstract` flag marks the field so that it needs to be implemented by a subschema, so using it just to make the test result look different is not going to work.

Kind regards,
Hans

**Reply by jsumners** [reply]
> Can you actually change your test to compare the validation result field by field, rather than comparing the whole table?

Ideally, no. The purpose was to write an `assert.same` test. I ended up writing an `assert.match` that does roughly what https://node-tap.org/docs/api/asserts/#tmatchfound-pattern-message-extra does for objects.
